<!-- This .launch file serves to perform the Circular, Eight-like, or
Target Reaching trajectories using the default factory-installed position
control mode of Baxter.
To select one of the trajectories, uncomment the code-block corresponding to
"CIRCLE TRAJECTORY", "EIGHT TRAJECTORY", or  "TARGET REACHING", and comment
the other two.-->

<launch>
<!-- GLOBAL PARAMETERS
  -Use_sim_time:
    False-Run in real time.
    True-Run in simulation time.
  -Reference_time: Initialization time. It is set up in edlut simulator node (not used in this case)-->
<param name="use_sim_time" value="false" type="bool" />
<param name="reference_time" value="0.0" type="double" />


<!-- ROBOT ENABLER NODE
It enables the robot. Once done it publishes a message advertising
the robot is ready and then the node is shutdown. -->
<node pkg="edlut_ros" name="robot_enabler" type="robot_enabler_node.py" output="screen">
</node>


<!-- TRAJECTORY FROM FILE GENERATOR NODE
It generates a trajectory reading it from a .txt file in the format:
J0 J1 J2 J3 J4 ...
J0 J1 J2 J3 J4 ...
.
.
J0 J1 J2 J3 J4 ...
Each line contains the position/velocity for each joint. Each line corresponds to
a sample of the trajectory.
This node reads the files and publishes the trajectory on the desired position
and desired velocity topics. A trajectory is generated for every joint (joint_list param).
The node also generates a control signal and avertises it (control_topic) to notify when the different learning trials start.
Total number of trials and their length (seconds) is also specified. -->
<!--  CIRCLE TRAJECTORY -->
<node pkg="edlut_ros" name="q_trajectory_generator" type="q_trajectory_generator_node" output="screen" clear_params="true">
  # Define private parameters for the EDLUT simulation
  <param name="number_of_trials" value="1" type="int" />
  <param name="trial_length" value="10000" type="double" />
  <param name="joint_state_topic" value="robot/joint_states" type="str" />
  <param name="update_frequency" value="500.0" type="double" />
  <param name="samples" value="1000" type="int" />
  <param name="trajectory_frequency" value="0.5" type="double" />
  <param name="positions_file_name" value="/home/baxterpi/raspberry_nodes/src/edlut_ros/circles/horizontalCircle_1000samples_2seconds_r12_position.txt" type="str" />
  <param name="velocities_file_name" value="/home/baxterpi/raspberry_nodes/src/edlut_ros/circles/horizontalCircle_1000samples_2seconds_r12_velocity.txt" type="str" />
    <rosparam param="joint_list">
      [left_s0,left_s1,left_e0,left_e1,left_w0,left_w1,left_w2]
    </rosparam>
    <rosparam param="max_pos_amplitude">
      [1.7,1.04, 3.05, 2.61, 3.05, 2.09, 3.05]
    </rosparam>
    <rosparam param="min_pos_amplitude">
      [-1.7, -2.14, -3.05, -0.05, -3.05, -1.57, -3.05]
    </rosparam>
    <rosparam param="max_vel_amplitude">
      [2.0, 2.0, 2.0, 2.0, 4.0, 4.0, 4.0]
    </rosparam>
    <rosparam param="min_vel_amplitude">
      [-2.0, -2.0, -2.0, -2.0, -4.0, -4.0, -4.0]
    </rosparam>
    <param name="desired_position_topic" value="edlut_ros/desired_position/compact" type="str" />
    <param name="desired_velocity_topic" value="edlut_ros/desired_velocity/compact" type="str" />
    <param name="control_topic" value="edlut_ros/in_trial" type="str" />
    <param name="joint_command_topic" value="robot/limb/left/joint_command" type="str" />
  </node>


<!-- BAXTER POSITION COMMAND
This node sends Baxter the position commands according to the desired
trajectory. The default factory-installed position control mode is used.-->
<node pkg="edlut_ros" name="torque_command_node" type="baxter_position_command_node" output="screen">
  # Define private parameters for the spike decoder
  <param name="input_topic" value="edlut_ros/desired_position/compact" type="str" />
  <param name="output_topic" value="robot/limb/left/joint_command" type="str" />
  <rosparam param="joint_list">
    [left_s0,left_s1,left_e0,left_e1,left_w0,left_w1,left_w2]
  </rosparam>
  <param name="limb" value="left" type="str" />
  <param name="control_topic" value="edlut_ros/in_trial" type="str" />
  <param name="sampling_frequency" value="500.0" type="double" />
  <param name="disable_gravity_compensation" value="false" type="bool" />
  <param name="gravity_topic" value="robot/limb/left/suppress_gravity_compensation" type="str" />
</node>


</launch>
